src/Lib.P2P/Routing/DhtService.cs

```diff
@@ -0,0 +1,444 @@
+            {
+                QueryType = MessageType.GetProviders,
+                QueryKey = id.Hash,
+                Dht = this,
+                AnswersNeeded = limit,
+            };
+            
+            if (action != null)
+            {
+                dQuery.AnswerObtained += (s, e) => action.Invoke(e);
+            }
+
+            // Add any providers that we already know about.
+            var providers = ContentRouter
+               .Get(id)
+               .Select(pid => pid == SwarmService.LocalPeer.Id
+                    ? SwarmService.LocalPeer
+                    : SwarmService.RegisterPeer(new Peer {Id = pid}));
+            
+            foreach (var provider in providers)
+            {
+                dQuery.AddAnswer(provider);
+            }
+
+            // Ask our peers for more providers.
+            if (limit > dQuery.Answers.Count())
+            {
+                await dQuery.RunAsync(cancel).ConfigureAwait(false);
+            }
+
+            return dQuery.Answers.Take(limit);
+        }
+
+        /// <summary>
+        ///   Advertise that we can provide the CID to the X closest peers
+        ///   of the CID.
+        /// </summary>
+        /// <param name="cid">
+        ///   The CID to advertise.ipfs
+        /// </param>
+        /// <remarks>
+        ///   This starts a background process to send the AddProvider message
+        ///   to the 4 closest peers to the <paramref name="cid"/>.
+        /// </remarks>
+        public void Advertise(Cid cid)
+        {
+            _ = Task.Run(async () =>
+            {
+                var advertsNeeded = 4;
+                var message = new DhtMessage
+                {
+                    Type = MessageType.AddProvider,
+                    Key = cid.Hash.ToArray(),
+                    ProviderPeers = new[]
+                    {
+                        new DhtPeerMessage
+                        {
+                            Id = SwarmService.LocalPeer.Id.ToArray(),
+                            Addresses = SwarmService.LocalPeer.Addresses
+                               .Select(a => a.WithoutPeerId().ToArray())
+                               .ToArray()
+                        }
+                    }
+                };
+                
+                var peers = RoutingTable
+                   .NearestPeers(cid.Hash)
+                   .Where(p => p != SwarmService.LocalPeer);
+
+                foreach (var peer in peers)
+                {
+                    try
+                    {
+                        await using (var stream = await SwarmService.DialAsync(peer, ToString()))
+                        {
+                            Serializer.SerializeWithLengthPrefix(stream, message, PrefixStyle.Base128);
+                            await stream.FlushAsync();
+                        }
+
+                        if (--advertsNeeded == 0)
+                        {
+                            break;
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Refactor the code in order to remove this break statement.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=5051852)